source/common/network/connection_balancer_impl.cc

```diff
@@ -0,0 +1,40 @@
+#include "common/network/connection_balancer_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+void ExactConnectionBalancerImpl::registerHandler(BalancedConnectionHandler& handler) {
+  absl::MutexLock lock(&lock_);
+  handlers_.push_back(&handler);
+}
+
+void ExactConnectionBalancerImpl::unregisterHandler(BalancedConnectionHandler& handler) {
+  absl::MutexLock lock(&lock_);
+  handlers_.erase(std::find(handlers_.begin(), handlers_.end(), &handler));
+}
+
+ConnectionBalancer::BalanceConnectionResult
+ExactConnectionBalancerImpl::balanceConnection(ConnectionSocketPtr&& socket,
+                                               BalancedConnectionHandler& current_handler) {
+  BalancedConnectionHandler* min_connection_handler = nullptr;
+  {
+    absl::MutexLock lock(&lock_);
+    for (const auto handler : handlers_) {
```

Comment:
prefer explicit type to auto here as it's not contextually obvious what exact type handlers_ contains. https://google.github.io/styleguide/cppguide.html#Type_deduction

Also I think you want a const ref? Oh ... now I from the assignment below it's a pointer so we don't need a ref to it.